spec/compiler/type_inference/instance_var_spec.cr

```diff
@@ -2157,6 +2157,154 @@ describe "Type inference: instance var" do
       )) { int32 }
   end
 
+  it "infers from class method that has type annotation" do
+    assert_type(%(
+      class Bar
+        def self.bar : Bar
+          Bar.new
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar.bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from class method that has type annotation, in generic class" do
+    assert_type(%(
+      class Bar
+        def self.bar : Bar
+          Bar.new
+        end
+      end
+
+      class Foo(T)
+        def initialize
+          @bar = Bar.bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo(Int32).new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from generic class method that has type annotation" do
+    assert_type(%(
+      class Bar(T)
+        def self.bar : self
+          Bar(T).new
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar(Int32).bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { (types["Bar"] as GenericClassType).instantiate([int32] of TypeVar) }
+  end
+
+  it "infers from generic class method that has type annotation, without instantiating" do
+    assert_type(%(
+      class Bar(T)
+        def self.bar : Int32
+          1
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar.bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { int32 }
+  end
+
+  it "infers from class method that has type annotation, with overload" do
+    assert_type(%(
+      class Baz
+      end
+
+      class Bar
+        def self.bar : Baz
+          Baz.new
+        end
+
+        def self.bar(x) : Bar
+          Bar.new
+        end
+
+        def self.bar(x) : Baz
+          yield
+          Bar.new
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar.bar(1)
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from class method that has type annotation, with multiple overloads matching, all with the same type" do
```

Comment:
What happens if return type of methods do not match?
Error or union types?
Does method resolution carry on argument type information at the moment of method resolution?